Add `.bot` and `.idx` files to EK60/EK80 parsing

ctuguinay opened the pull request:
Addresses #1315

ctuguinay commented:
@leewujung This should also be ready to look at. Only recent change I made was the `**kwarg` thing, Not sure why it still says read the docs pending. The test docs completed already.

leewujung commented on echopype/convert/api.py:
```diff
@@ -294,23 +308,48 @@ def _check_file(
+        bot_fsmap = fsspec.get_mapper(bot_file, **storage_options)
+        if not bot_fsmap.fs.exists(bot_fsmap.root):
+            raise FileNotFoundError(
+                f"There is no file named {bot_file}. The `.BOT` file must be contained in "
+                + " the same directory as that of the input 'raw' file."
```
This is super tiny but i think since the backtick is really used for docstring, maybe we can skip it here.
```suggestion
                f"There is no file named {bot_file}. The .BOT file must be contained in "
                + " the same directory as that of the input 'raw' file."
```

leewujung commented on echopype/convert/api.py:
```diff
@@ -294,23 +308,48 @@ def _check_file(
+        idx_fsmap = fsspec.get_mapper(idx_file, **storage_options)
+        if not idx_fsmap.fs.exists(idx_fsmap.root):
+            raise FileNotFoundError(
+                f"There is no file named {idx_file}. The `.IDX` file must be contained in "
+                + " the same directory as that of the input 'raw' file."
```
```suggestion
                f"There is no file named {idx_file}. The .IDX file must be contained in "
                + " the same directory as that of the input 'raw' file."
```

leewujung commented on echopype/convert/set_groups_base.py:
```diff
@@ -362,3 +362,141 @@ def beam_groups_to_convention(
+        platform_ds["time3"] = platform_ds["time3"].assign_attrs(
+            {
+                "axis": "T",
+                "long_name": "Timestamps from `.IDX` datagrams",
```
Same minor thing: remove all backticks from the comment here and below?
```suggestion
                "long_name": "Timestamps from .IDX datagrams",
```

leewujung commented on echopype/convert/set_groups_base.py:
```diff
@@ -362,3 +362,141 @@ def beam_groups_to_convention(
+        platform_ds["latitude_idx"] = platform_ds["latitude_idx"].assign_attrs(
+            {
+                "long_name": "Index File Derived Platform Latitude",
+                "comment": "Data derived from `.IDX` datagrams. Aligns time-wise with this "
+                + "dataset's `time3` dimension. This is different from `NMEA` latitude.",
```
```suggestion
                "comment": "Data derived from `.IDX` datagrams. Aligns time-wise with this "
                + "dataset's `time3` dimension. "
                + "This is different from latitude stored in the NMEA data gram.",
```

ctuguinay replied:
Ah yeah I agree. Good catch

leewujung commented on echopype/convert/set_groups_base.py:
```diff
@@ -362,3 +362,141 @@ def beam_groups_to_convention(
+        platform_ds["longitude_idx"] = platform_ds["longitude_idx"].assign_attrs(
+            {
+                "long_name": "Index File Derived Platform Longitude",
+                "comment": "Data derived from `.IDX` datagrams. Aligns time-wise with this "
+                + "dataset's `time3` dimension. This is different from `NMEA`longitude.",
```
```suggestion
                "comment": "Data from the IDX datagrams. Aligns time-wise with this "
                + "dataset's `time3` dimension. "
                + "This is different from longitude from the NMEA datagram.",
```

leewujung commented on echopype/convert/set_groups_base.py:
```diff
@@ -362,3 +362,141 @@ def beam_groups_to_convention(
+        )
+        vendor_ds["ping_time"] = vendor_ds["ping_time"].assign_attrs(
+            {
+                "long_name": "Timestamps from `.BOT` datagrams",
```
```suggestion
                "long_name": "Timestamps from the BOT datagrams",
```

leewujung commented on echopype/convert/set_groups_base.py:
```diff
@@ -362,3 +362,141 @@ def beam_groups_to_convention(
+            }
+        )
+        vendor_ds["detected_seafloor_depth"] = vendor_ds["detected_seafloor_depth"].assign_attrs(
+            {"long_name": "Echosounder detected seafloor depth from `.BOT` datagrams."}
```
```suggestion
            {"long_name": "Echosounder detected seafloor depth from the BOT datagrams."}
```

leewujung commented on echopype/convert/set_groups_ek60.py:
```diff
@@ -310,6 +310,16 @@ def set_platform(self) -> xr.Dataset:
         ds = ds.assign_attrs(platform_dict)
 
+        # If `.IDX` file exists and `.IDX` data is parsed
+        if (
+            (self.parser_obj.idx_file != "")
+            and self.parser_obj.idx["vessel_distance"]
+            and self.parser_obj.idx["latitude_idx"]
+            and self.parser_obj.idx["longitude_idx"]
+            and self.parser_obj.idx["timestamp"]
+        ):
```
I think this is fine, but I am curious about if the later checks (those involving `self.parser_obj.idx`) are included to catch cases where `self.parser_obj.idx_file != ""` is satisfied but `self.parser_obj.idx` does not contain any data?

ctuguinay replied:
Yup, that's what it's meant to catch. In the case where the parser reads nothing from the IDX file

leewujung commented on echopype/tests/convert/test_convert_azfp.py:
```diff
@@ -196,7 +196,7 @@ def test_convert_azfp_01a_pressure_temperature(azfp_path):
 def test_load_parse_azfp_xml(azfp_path):
     azfp_xml_path = azfp_path / '23081211.XML'
-    parseAZFP = ParseAZFP(None, str(azfp_xml_path))
+    parseAZFP = ParseAZFP(None, str(azfp_xml_path), {}, "")
```
Seems like this line is ok without the additions?

ctuguinay replied:
Ah yeah I removed this. Thanks!

leewujung commented on echopype/tests/convert/test_convert_ek.py:
```diff
@@ -0,0 +1,147 @@
+        ("echopype/test_data/ek80/idx_bot/Hake-D20230711-T182702.raw", "EK80"),
+    ]
+)
+def test_convert_ek60_with_bot_file(file, sonar_model):
```
```suggestion
def test_convert_ek_with_bot_file(file, sonar_model):
```

leewujung commented on echopype/tests/convert/test_convert_ek.py:
```diff
@@ -0,0 +1,147 @@
+        ("echopype/test_data/ek80/idx_bot/Hake-D20230711-T182702.raw", "EK80"),
+    ]
+)
+def test_convert_ek60_with_idx_file(file, sonar_model):
```
```suggestion
def test_convert_ek_with_idx_file(file, sonar_model):
```

leewujung reviewed (approved):
@ctuguinay : Thanks for the PRs, looks great! I only have cosmetic suggestions.

leewujung commented on echopype/convert/set_groups_base.py:
```diff
@@ -362,3 +362,143 @@ def beam_groups_to_convention(
+                "calendar": DEFAULT_TIME_ENCODING["calendar"],
+            },
+        )
+        platform_ds = platform_ds.assign(
+            {
+                "vessel_distance": xr.DataArray(
+                    np.array(self.parser_obj.idx["vessel_distance"]),
+                    dims=("time3"),
+                    coords={"time3": timestamp_array},
+                ),
+                "latitude_idx": xr.DataArray(
+                    np.array(self.parser_obj.idx["latitude_idx"]),
+                    dims=("time3"),
+                    coords={"time3": timestamp_array},
+                ),
+                "longitude_idx": xr.DataArray(
+                    np.array(self.parser_obj.idx["longitude_idx"]),
+                    dims=("time3"),
+                    coords={"time3": timestamp_array},
+                ),
+            }
+        )
```
Sorry I should have noticed this yesterday... I just checked the IDX0 datagram content. I think we should add `ping_number` and `filer_offset` to the stored data as well, since in the manual there are some intricacies in how the pings are associated:
https://www.simrad.online/ek80/interface/ek80_interface_en_a4.pdf#page=162

Also, to make it easy to recognize these, how about prefixing the original data variable names with `idx_`, so there is no confusion of which one is which? i.e. `idx_distance`, `idx_latitude`, etc.

ctuguinay commented:
Pinned `numpy<2` and made the IDX prefix change. I'll read up on the datagrams

ctuguinay commented:
Added ping number and file offset to values that are parsed in the datagram and added to the group. I still don't completely understand file offset...so I'll figure that out first before I add their corresponding group attributes. I've also never used the replay bar in Echoview so I'll try that out tomorrow.

leewujung commented:
I think for the replay thing they are referring to the EK80 software. The windows machine also has that installed.

ctuguinay commented:
@leewujung Ok this is ready for review again haha. Let's figure out #1344 another time, I left a small TODO in the code as a reminder for that.

leewujung commented on echopype/convert/set_groups_base.py:
```diff
@@ -396,18 +396,28 @@ def _add_index_data_to_platform_ds(
                     np.array(self.parser_obj.idx["vessel_distance"]),
                     dims=("time3"),
                     coords={"time3": timestamp_array},
                 ),
```
Just to make it absolutely clear where these came from
```suggestion
                "idx_ping_number": xr.DataArray(
                    np.array(self.parser_obj.idx["ping_number"]),
                    dims=("time3"),
                    coords={"time3": timestamp_array},
                ),
                "idx_file_offset": xr.DataArray(
                    np.array(self.parser_obj.idx["file_offset"]),
                    dims=("time3"),
                    coords={"time3": timestamp_array},
                ),
                "idx_vessel_distance": xr.DataArray(
                    np.array(self.parser_obj.idx["vessel_distance"]),
                    dims=("time3"),
                    coords={"time3": timestamp_array},
                ),
```

leewujung replied:
And just because I am looking at this... why wouldn't you do the attribute assignment here in each of the xr.DataArray too?

ctuguinay replied:
Ah yeah you're right I'll do that

leewujung commented on echopype/convert/parse_base.py:
```diff
@@ -563,9 +563,11 @@ def _read_datagrams(self, fid):
                 self.idx["vessel_distance"].append(new_datagram["distance"])
-                self.idx["latitude_idx"].append(new_datagram["latitude"])
-                self.idx["longitude_idx"].append(new_datagram["longitude"])
+                self.idx["idx_latitude"].append(new_datagram["latitude"])
+                self.idx["idx_longitude"].append(new_datagram["longitude"])
```
Since these are already under `self.idx` you could remove the idx_ here also.

ctuguinay replied:
made this change 👍

leewujung reviewed (dismissed):
@ctuguinay : The changes look great! I only have very nitpicking comments. Feel free to merge!

Also I wonder what the difference between the 2 EK60 test files (or between the 2 test EK80 files) are?

ctuguinay commented:
> Also I wonder what the difference between the 2 EK60 test files (or between the 2 test EK80 files) are?

No big difference. I just wanted to make sure I had two of each, since I like the number 2 😄

ctuguinay commented:
@leewujung 😭 ok it now requires another review approval before I can merge since I made the changes you suggested above
![image](https://github.com/OSOceanAcoustics/echopype/assets/87830138/3d2038d4-6427-4528-9737-aea1abc345a5)

leewujung reviewed (approved):
🤯 ok here you go!
